File: dailystreak/LC-1861(6thMay26)/optimal.java
//time complexity - o(m*n)
//space complexity - o(m*n) because of rotating list 
class Solution {
    public char[][] rotateTheBox(char[][] boxGrid) {

        int m = boxGrid.length;
        int n = boxGrid[0].length;

        // Move stones towards right due to gravity
        for (int i = 0; i < m; i++) {

            // position where next stone should go
            int empty = n - 1;

            // traverse from right to left
            for (int j = n - 1; j >= 0; j--) {

                // obstacle found
                if (boxGrid[i][j] == '*') {

                    empty = j - 1;
                }

                // stone found
                else if (boxGrid[i][j] == '#') {

                    // move stone to empty position
                    char temp = boxGrid[i][empty];
                    boxGrid[i][empty] = '#';
                    boxGrid[i][j] = temp;

                    // next empty position shifts left
                    empty--;
                }
            }
        }

        // Rotate matrix m*n - n*m
        char[][] ans = new char[n][m];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {

                ans[j][m - 1 - i] = boxGrid[i][j];
            }
        }

        return ans;
    }
}

File: dailystreak/LC-2657(20thMay26)/optimal.py
#time complexity: O(n)
#space complexity: O(n)

class Solution:
    def findThePrefixCommonArray(self, A: List[int], B: List[int]) -> List[int]:
        
        n = len(A)

        visited = [0] * (n + 1)

        ans = []
        common = 0

        for i in range(n):

            visited[A[i]] += 1
            if visited[A[i]] == 2:
                common += 1

            visited[B[i]] += 1
            if visited[B[i]] == 2:
                common += 1

            ans.append(common)

        return ans